src/java.base/share/classes/jdk/internal/foreign/AbstractMemorySegmentImpl.java

```diff
@@ -469,6 +469,10 @@ public int hashCode() {
 
     public static AbstractMemorySegmentImpl ofBuffer(Buffer bb) {
         Objects.requireNonNull(bb);
+        // StringCharBuffer is packet-private
+        if ("java.nio.StringCharBuffer".equals(bb.getClass().getName())) {
```

Comment:
IMHO would be better to call Buffer::isDirect, and then check if Buffer::array is == null.

Reply:
Nevermind, the array is not given away for read-only buffers... grrr

Reply:
We get the array in the code below. I guess we can check if it's `null` there?

Reply:
yep - that works as `NIO_ACCESS.getBufferBase` goes straight to `Buffer.hb`, which is what we need.